File: Wordfolio.Frontend/src/types/entry.ts
export type DefinitionSource = "Api" | "Manual";
export type TranslationSource = "Api" | "Manual";
export type ExampleSource = "Api" | "Custom";

export interface Example {
    readonly id: number;
    readonly exampleText: string;
    readonly source: ExampleSource;
}

export interface Definition {
    readonly id: number;
    readonly definitionText: string;
    readonly source: DefinitionSource;
    readonly displayOrder: number;
    readonly examples: Example[];
}

export interface Translation {
    readonly id: number;
    readonly translationText: string;
    readonly source: TranslationSource;
    readonly displayOrder: number;
    readonly examples: Example[];
}

export interface Entry {
    readonly id: number;
    readonly vocabularyId: number;
    readonly entryText: string;
    readonly createdAt: Date;
    readonly updatedAt: Date | null;
    readonly definitions: Definition[];
    readonly translations: Translation[];
}

File: Wordfolio.Api/Wordfolio.Api.Domain/Entries/Entry.fs
namespace Wordfolio.Api.Domain.Entries

open System

open Wordfolio.Api.Domain
open Wordfolio.Api.Domain.Shared

type DefinitionSource =
    | Api
    | Manual

type TranslationSource =
    | Api
    | Manual

type ExampleSource =
    | Api
    | Custom

type Example =
    { Id: ExampleId
      ExampleText: string
      Source: ExampleSource }

type Definition =
    { Id: DefinitionId
      DefinitionText: string
      Source: DefinitionSource
      DisplayOrder: int
      Examples: Example list }

type Translation =
    { Id: TranslationId
      TranslationText: string
      Source: TranslationSource
      DisplayOrder: int
      Examples: Example list }

type Entry =
    { Id: EntryId
      VocabularyId: VocabularyId
      EntryText: string
      CreatedAt: DateTimeOffset
      UpdatedAt: DateTimeOffset option
      Definitions: Definition list
      Translations: Translation list }

type ExampleInput =
    { ExampleText: string
      Source: ExampleSource }

type DefinitionInput =
    { DefinitionText: string
      Source: DefinitionSource
      Examples: ExampleInput list }

type TranslationInput =
    { TranslationText: string
      Source: TranslationSource
      Examples: ExampleInput list }

type CreateEntryParameters =
    { UserId: UserId
      VocabularyId: VocabularyId option
      EntryText: string
      Definitions: DefinitionInput list
      Translations: TranslationInput list
      AllowDuplicate: bool
      CreatedAt: DateTimeOffset }

type DraftsVocabularyData =
    { Vocabulary: Vocabulary
      Entries: Entry list }
